### PR TITLE
Adding ubuntu 16.04 as base image.

### LINKED ISSUE #64: Add ubuntu 16.04 base image
Just like [ubuntu 18.04](https://github.com/dlespiau/footloose/tree/master/images/ubuntu18.04), we should add the 16.04 LTS base image to footloose.

### PR DESCRIPTION
**Pull request by Skarlso**
Closes #64.

### CODE REVIEW

**Comment by Skarlso** [comment]
@dlespiau you need to push up that image I believe?

**Comment by Skarlso** [comment]
Building the container did succeed. Although this part I had to remove:

```
RUN systemctl disable \
      networkd-dispatcher.service
```

Success: 
```bash
Successfully built e0dc65e274b0
Successfully tagged skarlso/ubuntu16:latest
```

**Comment by dlespiau** [comment]
Yes, I'd need to push the imagine by hand, sorry about that. Have you managed to run an ubuntu 16.04 container successfully locally? I failed to do so in PR #65.

**Comment by dlespiau** [comment]
I removed networkd-dispatcher.service because I didn't feel it was something necessary, the containers are connected to a linux bridge and seem unlikely to have to handle network going away, especially because another process has to listen to that signal over d-bus.

But maybe that's fine and we should add it back :man_shrugging:  :)

**Comment by Skarlso** [comment]
@dlespiau Run as in create one? Yeah, the output is there of that run. Or do you mean with footloose?

I can try that too by adjusting the repository URL I guess. :)

**Comment by dlespiau** [comment]
Yes, you can try it locally, just need to have the config file with the right image name/tag by either editing `footloose.yaml` or using the `--image` option of create. When I did that, the Ubuntu 16.06 image did not start sshd, so that would need debugging.

**Comment by Skarlso** [comment]
I'll check it out as soon as I can. :) 👍

**Comment by Skarlso** [comment]
Oh GODs this sucks massively. Ubuntu 16 and systemd aren't working together AT ALL.

That said, sshd also isn't working with sbin/init. it's simply not starting, as you said. But there is a bigger problem than that. You can launch a sshd service with a container running it as a daemon if your entrypoint isn't sbin, but rahter the SSH command as a daemon like `/usr/sbin/sshd -D`. This runs correctly and allows for an SSH server to persist, however, you guess it, this means that on ubuntu16 you'd have to run a different command.

Here is where supervisord comes into play. It's handling running SSH so commands can be executed or ssh works once it's started. But this is all bad now, because supervisord needs to be the one running instead of init. :/

File: pkg/cluster/cluster.go [hunk 1]
```diff
@@ -140,7 +140,7 @@ func (c *Cluster) createMachine(machine *Machine, i int) error {
 
 	_, err := docker.Run(machine.spec.Image,
 		runArgs,
-		[]string{"/sbin/init"},
+		[]string{"/usr/bin/supervisord"},
```
Comment (Skarlso) [inline pkg/cluster/cluster.go]: @dlespiau Without this, SSH won't start. :/

File: pkg/cluster/cluster.go [hunk 2]
```diff
@@ -165,11 +165,11 @@ func (c *Cluster) createMachineRunArgs(machine *Machine, name string, i int) []s
 	runArgs := []string{
 		"-it", "-d", "--rm",
 		"--name", name,
-		"--hostname", machine.Hostname(),
```
Comment (Skarlso) [inline pkg/cluster/cluster.go]: @dlespiau Now this is more interesting. One of these f*cks up SSH :D No idea yet which one, probably the hostname. without these ssh was unable to start exiting with 255. I run out of time to investigate this further.

**Comment by Skarlso** [comment]
That said ((the above )) this now actually works:

```
~/goprojects/footloose issue_64_ubuntu_16_04* [AWS:default]
❯ ./footloose ssh root@node0
Welcome to Ubuntu 16.04.5 LTS (GNU/Linux 4.9.125-linuxkit x86_64)

 * Documentation:  https://help.ubuntu.com
 * Management:     https://landscape.canonical.com
 * Support:        https://ubuntu.com/advantage

The programs included with the Ubuntu system are free software;
the exact distribution terms for each program are described in the
individual files in /usr/share/doc/*/copyright.

Ubuntu comes with ABSOLUTELY NO WARRANTY, to the extent permitted by
applicable law.

root@0c00f6681b59:~# exit
```

We need to find out how to handle this, or either not support 16.04 at all. Your call.

**Comment by Skarlso** [comment]
I have a solution which isn't a hack.

Make that init run in the cluster a config value. Default to sbin/init and make it overriddable by config. Choosing Ubuntu 16.04 one can override it to run supervisord instead. The rest would just work in theory.

**Comment by Skarlso** [comment]
```
❯ ./footloose ssh root@node0 -- echo "success"
success
```

**Comment by Skarlso** [comment]
@dlespiau Okay, so I figured out which setting kills SSH and it's probably because it over maps the running processes which I don't understand the reason for. :) 

Could you please explain this one: 
```
// "--tmpfs", "/run",
```

Why tmpfs /run? :)

**Comment by Skarlso** [comment]
huh actually apparently /var/run/sshd is missing. :O because it's somehow now a symlink to /run/lock. :O WTH?

**Comment by Skarlso** [comment]
Oh.. my bad. Totally forgot that linux has now var/run symlinked for a while now.

**Comment by Skarlso** [comment]
Nope. Even if I create /run/sshd the tmpfs /run totally overrides it.... Unless I create the sshd folder it can't start the service since it's privilege folder, `/var/run/sshd` is missing.

**Comment by Skarlso** [comment]
@dlespiau Everything works. :)))))))))

Sorry, for the noise. Now supervisord is creating the folder SSHD needs first. After that all else is working fine. :) This is the configuration example for a working ubuntu 16:

```
cluster:
  name: cluster
  privateKey: cluster-key
machines:
- count: 1
  spec:
    image: docker.io/skarlso/footloose:ubuntu16
    name: node%d
    portMappings:
    - containerPort: 22
    cmd: /usr/bin/supervisord
```

What do you think? :)

**Comment by dlespiau** [comment]
I'll have to look a bit more into this one, I will have time for it a bit later this week :)

**Comment by Skarlso** [comment]
@dlespiau Hey man. :) you require any info from me on this one? :)

**Comment by dlespiau** [comment]
`master` is now testing that the images have a set of basic command. You need to add ` net-tools iproute2 iputils-ping curl wget vim-tiny` to the list of packages to install to get the tests to pass. The CI scripts are now building images before running the tests, so the `Dockerfile` changes should be picked up!

**Comment by Skarlso** [comment]
Cool! I'll do that.

**Comment by Skarlso** [comment]
@dlespiau Oh really nice. Tests passed. :) Lovely!

**Comment by Skarlso** [comment]
Oh... I did not expect this to fail... :/ going to investigate.

**Comment by dlespiau** [comment]
We're down to the docker-in-docker test not passing.

```
Processing triggers for dbus (1.10.6-1ubuntu3.3) ...
Failed to connect to bus: No such file or directory
time="2019-03-18T13:42:58Z" level=fatal msg="exit status 1"
```

I'm still a bit puzzled by not being able to use systemd with ubuntu 16.04, a quick search around seems to suggest it should be possible, but don't really have time to investigate right now. I can propose a way forward:

- Remove the docker-in-docker test for ubuntu 16.04, tests should pass then
- Merge this PR
- Open an issue about investigating systemd for ubuntu 16.04
- Open an issue about investigating docker-in-docker for ubuntu 16.04

**Comment by Skarlso** [comment]
@dlespiau I tried the systemd approach but there were problems regarding it not being able to start or switch to systemd when the container comes on. Problems relating to dbus not found and systemd not running at all. Init wasn't able to properly start systemd in the end and I had to resort to a different medium running the SSH daemon on start. Which is supervisord.

**Comment by Skarlso** [comment]
Huh. Interesting. This should have worked. It's not picking up the `cmd` from the config? 🤔

**Comment by dlespiau** [comment]
It is starting supervisord, see: https://travis-ci.org/weaveworks/footloose/builds/507854258#L3759

My guess would be that dbus isn't started and we can't rely on the socket activation mechanism from systemd.

**Comment by Skarlso** [comment]
This is the only project I saw that had systemd somewhat working. This also required a special treatment for docker host and some hacky things to happen inside the container too and run in `--privileged` mode. https://github.com/solita/docker-systemd

However, this is also unmaintained now, as are all the other projects trying to get systemd to run in an ubuntu 16 container. :D

also, this means that the test which is calling `systemctl start docker` will of course, not work. :/

**Comment by Skarlso** [comment]
BTW.... If I remove the `cmd` section... it is actually working.

```
INFO[0000] Deleting machine: test-docker-in-docker-ubuntu16.04-node0 ...
PASS
ok  	github.com/weaveworks/footloose/tests	34.760s
```
🤔

**Comment by dlespiau** [comment]
Installing docker.io pulls dbus in the image, so that may be why the docker-in-docker test works without cmd.

**Comment by Skarlso** [comment]
Unfortunately I'm already doing that here: https://github.com/weaveworks/footloose/pull/86/files#diff-dbfb59ddc0e28997ce347e0e8042b7a8R16
:/

**Comment by dlespiau** [comment]
ah well :)

**Comment by Skarlso** [comment]
I removed the docker in docker test for now. :)

**Comment by Skarlso** [comment]
Let me try something else...

**Comment by Skarlso** [comment]
Oh! Got systemctl working.

```
❯ docker exec -it 2230d916baa9 bash
root@node0:/# systemctl
UNIT                               LOAD   ACTIVE     SUB       DESCRIPTION
dev-sda1.device                    loaded activating tentative /dev/sda1
-.mount                            loaded active     mounted   /
dev-mqueue.mount                   loaded active     mounted   POSIX Message Queue File System
etc-hostname.mount                 loaded active     mounted   /etc/hostname
etc-hosts.mount                    loaded active     mounted   /etc/hosts
etc-resolv.conf.mount              loaded active     mounted   /etc/resolv.conf
tmp.mount                          loaded active     mounted   /tmp
init.scope                         loaded active     running   System and Service Manager
ondemand.service                   loaded active     running   LSB: Set the CPU Frequency Scaling governor to "ondemand"
rc-local.service                   loaded active     exited    /etc/rc.local Compatibility
systemd-journald.service           loaded active     running   Journal Service
systemd-tmpfiles-setup-dev.service loaded active     exited    Create Static Device Nodes in /dev
systemd-tmpfiles-setup.service     loaded active     exited    Create Volatile Files and Directories
-.slice                            loaded active     active    Root Slice
system.slice                       loaded active     active    System Slice
dbus.socket                        loaded active     listening D-Bus System Message Bus Socket
systemd-journald-audit.socket      loaded active     running   Journal Audit Socket
systemd-journald-dev-log.socket    loaded active     running   Journal Socket (/dev/log)
systemd-journald.socket            loaded active     running   Journal Socket
basic.target                       loaded active     active    Basic System
graphical.target                   loaded active     active    Graphical Interface
local-fs.target                    loaded active     active    Local File Systems
multi-user.target                  loaded active     active    Multi-User System
paths.target                       loaded active     active    Paths
slices.target                      loaded active     active    Slices
sockets.target                     loaded active     active    Sockets
swap.target                        loaded active     active    Swap
sysinit.target                     loaded active     active    System Initialization
timers.target                      loaded active     active    Timers
systemd-tmpfiles-clean.timer       loaded active     waiting   Daily Cleanup of Temporary Directories

LOAD   = Reflects whether the unit definition was properly loaded.
ACTIVE = The high-level unit activation state, i.e. generalization of SUB.
SUB    = The low-level unit activation state, values depend on unit type.

30 loaded units listed. Pass --all to see loaded but inactive units, too.
To show all installed unit files use 'systemctl list-unit-files'.
root@node0:/#
```

Now only thing I need is SSH to work.

**Comment by dlespiau** [comment]
Smells like progress :) :+1:

**Comment by Skarlso** [comment]
Dude! I made it work. You won't believe the rabbit hole  I got into. :D

**Comment by dlespiau** [comment]
tell me, tell me :D

**Comment by Skarlso** [comment]
Apparently... There is a bug. :D right now, there appears to be a configuration that is created in error or being accepted in error under /usr/lib/tmpfiles.d/sshd.conf.

I had a workaround by changing `/var/run/sshd` to `/run/ssdh` since now /var/run is a symlink.

However, it's STILL looking for that folder no matter how you change that file. :/

Long story short, apparently this will never be "fixed" because all they are saying is that you should upgrade your kernel and some such making systemd break. :(

Also, the ssh service was not enabled ( as a nice plus :D ).

All in all I tested things and now supervisor is not needed and init can be the main script that runs.

I'm leaving in the facility to override the executing command though. If you think that's not needed I can revert that part of the change.

The links are mostly: 
https://askubuntu.com/questions/1110828/ssh-failed-to-start-missing-privilege-separation-directory-var-run-sshd
https://bugs.launchpad.net/ubuntu/+source/systemd/+bug/1811580

I left these as a comment in the Dockerfile. :)

**Comment by Skarlso** [comment]
Oh yeah, also, I've put the test back since they should now be running fine. :)

**Comment by dlespiau** [comment]
Thanks for your perseverance looking into this, quite happy we can use systemd after all :) let's see what CI has to say about it :p

File: images/ubuntu16.04/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+# user locale env variables or we get warnings such as:
+#  bash: warning: setlocale: LC_ALL: cannot change locale
+RUN sed -i -e 's/^AcceptEnv LANG LC_\*$/#AcceptEnv LANG LC_*/' /etc/ssh/sshd_config
+
+RUN systemctl enable ssh.service
+
+# https://www.freedesktop.org/wiki/Software/systemd/ContainerInterface/
+STOPSIGNAL SIGRTMIN+3
+
+VOLUME [ "/sys/fs/cgroup" ]
```
Comment (dlespiau) [inline images/ubuntu16.04/Dockerfile]: This shouldn't be necessary, I think?

Comment (Skarlso) [reply images/ubuntu16.04/Dockerfile]: right you are. :)

**Comment by Skarlso** [comment]
Thanks. :) This was a very good lesson for me, to never settle for the first solution. :D

**Comment by dlespiau** [comment]
Tests have passed! somehow travis hasn't reported it back to github yet. May just merge if the event doesn't arrive in the next 5 minutes.

**Comment by Skarlso** [comment]
Oh, interesting. Yeah, I can't find anything else running. Odd.

Awesome!! :) \o/

**Comment by dlespiau** [comment]
Alright, merging! thanks for your awesome work here!

**Comment by Skarlso** [comment]
Yw. :) I had fun doing this :) Glad I could manage it. :)

**Pull request merged**